### PR TITLE
These are all the changes I want to commit in one merge

### PR DESCRIPTION
**Pull request by jwatte**
I removed the inner class Ref::Hold.
I've tested these changes on the samples, and the fix bugs, improve efficiency, and improve debuggability.
For reasons of edit history, the Ref::Hold commit is in here, and a later commit reverses it out.

### CODE REVIEW

File: gameplay/src/Container.cpp [hunk 1]
```diff
@@ -67,6 +67,7 @@ Container::~Container()
     std::vector<Control*>::iterator it;
     for (it = _controls.begin(); it < _controls.end(); it++)
     {
+        (*it)->_parent = nullptr;
```
Comment (seanpaultaylor) [inline gameplay/src/Container.cpp]: Please change to = NULL;

File: gameplay/src/TextBox.h [hunk 2]
```diff
@@ -112,6 +112,11 @@ class TextBox : public Label
 
     virtual void addListener(Control::Listener* listener, int eventFlags);
 
+    /**
+     * Update the text being edited.
+     */
+    void setText(char const *text) override;
```
Comment (seanpaultaylor) [inline gameplay/src/TextBox.h]: Please change to virtual void setText(...) and remove overriede.

File: gameplay/src/Transform.cpp [hunk 3]
```diff
@@ -99,43 +99,27 @@ const char* Transform::getTypeName() const
...
     {
         if (!isStatic())
         {
-            bool hasTranslation = !_translation.isZero();
             bool hasScale = !_scale.isOne();
             bool hasRotation = !_rotation.isIdentity();
 
             // Compose the matrix in TRS order since we use column-major matrices with column vectors and
             // multiply M*v (as opposed to XNA and DirectX that use row-major matrices with row vectors and multiply v*M).
-            if (hasTranslation || (_matrixDirtyBits & DIRTY_TRANSLATION) == DIRTY_TRANSLATION)
```
Comment (seanpaultaylor) [inline gameplay/src/Transform.cpp]: Why would we be removing a check for this?

File: gameplay/src/Font.cpp [hunk 4]
```diff
@@ -1371,9 +1371,11 @@ int Font::getIndexOrLocation(const char* text, const Rectangle& area, unsigned i
                 }
 
                 // Check against inLocation.
+                //  Note: g.width is smaller than g.advance, so if I only check against g.width, I will 
+                //  miss locations towards the right of the character.
                 if (destIndex == (int)charIndex ||
                     (destIndex == -1 &&
-                    inLocation.x >= xPos && inLocation.x < floor(xPos + g.width*scale + spacing) &&
+                    inLocation.x >= xPos && inLocation.x < floor(xPos + g.advance*scale + spacing) &&
```
Comment (seanpaultaylor) [inline gameplay/src/Font.cpp]: Good catch. Clearly a mistake in the code.

**Comment by seanpaultaylor** [comment]
Would you like to become a contributor ?

File: gameplay/src/Container.cpp (hunk 1 above)
Comment (jwatte) [inline gameplay/src/Container.cpp]: What version of compilers are you targeting? nullptr has been around for quite a while.

File: gameplay/src/TextBox.h (hunk 2 above)
Comment (jwatte) [inline gameplay/src/TextBox.h]: Before I make that change, which clearly makes the code more error prone, I'd like to understand why you suggest this. "override" on overrides is a fantastic error catcher. "virtual" on things you intend to override is redundant and, without override, makes for easy mistakes. And "override" has been around in gcc, clang, and MSVC for a long time!

File: gameplay/src/Transform.cpp (hunk 3 above)
Comment (jwatte) [inline gameplay/src/Transform.cpp]: The code is quite convoluted, for no real gain. Almost all real-life matrices actually have translation. And the additional branches are hard on CPU instruction scheduling. The real FPU cost comes in the rotation case, which makes that branch worth it. The rest of the code really just ends up paying for filling the cache. Thus, in practice, this code is simpler, and runs faster, on most architectures (including ARM 8.) But I'm mostly after the simplicity :-)

**Comment by jwatte** [comment]
Thanks, but no thanks at this time.




Sincerely,

Jon Watte


--
"I find that the harder I work, the more luck I seem to have." -- Thomas
Jefferson

File: gameplay/src/TextBox.h (hunk 2 above)
Comment (seanpaultaylor) [inline gameplay/src/TextBox.h]: I agree with the benefits. This engine started long before C++ 11. However, we are looking for also consistency. Throwing random qualifier is not something I am about to do. If you really want this benefit then add a feature issue for changing all virtual usage to override and fix it. Don't propose changes in a single API to break consistency for others using it.

**Comment by seanpaultaylor** [comment]
Np. I only need contributions from those willing. I will consider you pull requests the next time I am working in those files.

File: gameplay/src/Container.cpp (hunk 1 above)
Comment (seanpaultaylor) [inline gameplay/src/Container.cpp]: Looking for consistency here. I can go a change all the NULL to nullptr.

**Pull request merged**